client/utils/src/pubsub.rs

```diff
@@ -0,0 +1,263 @@
+{
+	fn drop(&mut self) {
+		if let Some(shared) = self.shared.upgrade() {
+			shared.lock().borrow_mut().unsubscribe(self.subs_id);
+		}
+	}
+}
+
+impl<M, R> Hub<M, R> {
+	/// Create a new instance of Hub (with default value for the Registry).
+	pub fn new(tracing_key: &'static str) -> Self
+	where
+		R: Default,
+	{
+		Self::new_with_registry(tracing_key, Default::default())
+	}
+
+	/// Create a new instance of Hub over the initialized Registry.
+	pub fn new_with_registry(tracing_key: &'static str, registry: R) -> Self {
+		let shared =
+			Shared { registry, sinks: Default::default(), id_sequence: Default::default() };
+		let shared = Arc::new(ReentrantMutex::new(RefCell::new(shared)));
+		Self { tracing_key, shared }
+	}
+
+	/// Subscribe to this Hub using the `subs_key: K`.
+	///
+	/// A subscription with a key `K` is possible if the Registry implements `Subscribe<K>`.
+	pub fn subscribe<K>(&self, subs_key: K) -> Receiver<M, R>
+	where
+		R: Subscribe<K> + Unsubscribe,
+	{
+		let shared_locked = self.shared.lock();
+		let mut shared_borrowed = shared_locked.borrow_mut();
+
+		let subs_id = shared_borrowed.id_sequence.next_id();
+
+		// The order (registry.subscribe then sinks.insert) is important here:
+		// assuming that `Subscribe<K>::subscribe` can panic, it is better to at least
+		// have the sink disposed.
+		shared_borrowed.registry.subscribe(subs_key, subs_id);
+
+		let (tx, rx) = crate::mpsc::tracing_unbounded(self.tracing_key);
+		assert!(shared_borrowed.sinks.insert(subs_id, tx).is_none(), "Used IDSequence to create another ID. Should be unique until u64 is overflowed. Should be unique.");
+
+		Receiver { shared: Arc::downgrade(&self.shared), subs_id, rx }
+	}
+
+	/// Send the message produced with `Trigger`.
+	///
+	/// This is possible if the registry implements `Dispatch<Trigger, Item = M>`.
+	pub fn send<Trigger>(&self, trigger: Trigger) -> <R as Dispatch<Trigger>>::Ret
+	where
+		R: Dispatch<Trigger, Item = M>,
+	{
+		let shared_locked = self.shared.lock();
+		let mut shared_borrowed = shared_locked.borrow_mut();
+		let (registry, sinks) = shared_borrowed.get_mut();
+
+		let dispatch_result = registry.dispatch(trigger, |subs_id, item| {
+			if let Some(tx) = sinks.get_mut(&subs_id) {
+				if let Err(send_err) = tx.unbounded_send(item) {
+					log::warn!("Sink with SubsID = {} failed to perform unbounded_send: {} ({} as Dispatch<{}, Item = {}>::dispatch(...))", subs_id, send_err, std::any::type_name::<R>(),
+					std::any::type_name::<Trigger>(),
+					std::any::type_name::<M>());
+				}
+			} else {
+				log::warn!(
+					"No Sink for SubsID = {} ({} as Dispatch<{}, Item = {}>::dispatch(...))",
+					subs_id,
+					std::any::type_name::<R>(),
+					std::any::type_name::<Trigger>(),
+					std::any::type_name::<M>(),
+				);
+			}
+		});
+
+		dispatch_result
+	}
+}
+
+impl<M, R> Shared<M, R> {
+	fn get_mut(&mut self) -> (&mut R, &mut HashMap<SeqID, TracingUnboundedSender<M>>) {
+		(&mut self.registry, &mut self.sinks)
+	}
```

Comment:
Any particular reason why you had to add this method? Seems a little unnecessary when you can just access the fields directly (and the method's private anyway, so it's not like this is used to expose it externally). Am I missing something here?